inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java

```diff
@@ -0,0 +1,196 @@
+    {
+
+    }
+
+    @Override
+    public void setModel(Object aModel)
+    {
+        if (aModel instanceof Set) {
+            nerAnnotations = (Set<AnnotationObject>) aModel;
+        }
+        else {
+            log.error("Expected model type: Set<TokenObject> - but was: [{}]",
+                aModel != null ? aModel.getClass() : aModel);
+        }
+    }
+
+    @Override
+    public void setUser(User user)
+    {
+        this.user = user;
+    }
+
+    @Override
+    public void setProject(Project project)
+    {
+        this.project = project;
+    }
+
+    /**
+     *
+     * @param inputData
+     *            All sentences to predict annotations for.
+     * @param <T>
+     * @return Predicted sentence.
+     *         Outer list: sentence level
+     *         Middle list: word level
+     *         Inner list: token level (predictions for each token)
+     */
+    @Override
+    public <T extends TokenObject> List<List<List<AnnotationObject>>> predictSentences(
+        List<List<T>> inputData)
+    {
+        List<List<List<AnnotationObject>>> result = new ArrayList<>();
+
+        for (List<T> sentence : inputData) {
+            List<List<AnnotationObject>> annotatedSentence = new ArrayList<>();
+            int sentenceIndex = 0;
+            while (sentenceIndex < sentence.size() - 1) {
+                TokenObject token = sentence.get(sentenceIndex);
+                List<AnnotationObject> word;
+
+                if (isNamedEntity(token)) {
+                    StringBuilder coveredText = new StringBuilder(token.getCoveredText());
+                    int endCharacter = token.getOffset().getEndCharacter();
+                    int endToken = token.getOffset().getEndToken();
+
+                    TokenObject nextTokenObject = sentence.get(sentenceIndex + 1);
+                    while (isNamedEntity(nextTokenObject)) {
+                        coveredText.append(" ").append(nextTokenObject.getCoveredText());
+                        endCharacter = nextTokenObject.getOffset().getEndCharacter();
+                        endToken = nextTokenObject.getOffset().getEndToken();
+                        sentenceIndex++;
+                        nextTokenObject = sentence.get(sentenceIndex + 1);
+                    }
+
+                    token.setCoveredText(coveredText.toString());
+                    token.setOffset(new Offset(token.getOffset().getBeginCharacter(), endCharacter,
+                        token.getOffset().getBeginToken(), endToken));
+                    word = predictToken(token);
+                    annotatedSentence.add(word);
+                }
+                sentenceIndex++;
+            }
+            result.add(annotatedSentence);
+        }
+        return result;
+    }
+
+    private List<AnnotationObject> predictToken(TokenObject token)
+    {
+        List<KBHandle> handles = new ArrayList<>();
+        for (KnowledgeBase kb : kbService.getKnowledgeBases(project)) {
```

Comment:
This should respect the concept feature traits: kb and scope - just like in the ConceptFeatureEditor.

Reply:
Adding ConceptFeatureTraits to inception-concept-linking will introduce a circular dependency to inception-ui-kb. I will create a new issue for that.